### ISSUE TITLE
Last closing round bracket is stripped out from the href

### ISSUE THREAD

**Issue by teimurjan** [markedjs/marked#1459]
**Describe the bug**

The last `)` is stripped out from the `href` when rendering HTML.

It works fine on CommonMark.

**To Reproduce**

`marked` receives the following content: `testText https://test.test?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:'2015-10-11T11:08:59.548Z',mode:absolute,to:'2015-10-11T11:23:59.548Z'))&_a=(columns:!(URL),index:logs_web_error,interval:auto,,sort:!(TimeStamp,desc))`. Instead of having the anchor element with this whole link I get:

```
<a href="https://test.test?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:'2015-10-11T11:08:59.548Z',mode:absolute,to:'2015-10-11T11:23:59.548Z'))&_a=(columns:!(URL),index:logs_web_error,interval:auto,,sort:!(TimeStamp,desc)">https://test.test?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:'2015-10-11T11:08:59.548Z',mode:absolute,to:'2015-10-11T11:23:59.548Z'))&_a=(columns:!(URL),index:logs_web_error,interval:auto,,sort:!(TimeStamp,desc)</a>)</p>
```

The last `)` is lost from the href somehow. 

1. [Marked Demo]( https://marked.js.org/demo/?outputType=html&text=https%3A%2F%2Ftest.test%3F_g%3D(refreshInterval%3A(display%3AOff%2Cpause%3A!f%2Cvalue%3A0)%2Ctime%3A(from%3A%272015-10-11T11%3A08%3A59.548Z%27%2Cmode%3Aabsolute%2Cto%3A%272015-10-11T11%3A23%3A59.548Z%27))%26_a%3D(columns%3A!(URL)%2Cindex%3Alogs_web_error%2Cinterval%3Aauto%2C%2Csort%3A!(TimeStamp%2Cdesc))&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tables%22%3A%20true%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)
2. [CommonMark Demo](https://spec.commonmark.org/dingus/?text=https%3A%2F%2Ftest.test%3F_g%3D(refreshInterval%3A(display%3AOff%2Cpause%3A!f%2Cvalue%3A0)%2Ctime%3A(from%3A%272015-10-11T11%3A08%3A59.548Z%27%2Cmode%3Aabsolute%2Cto%3A%272015-10-11T11%3A23%3A59.548Z%27))%26_a%3D(columns%3A!(URL)%2Cindex%3Alogs_web_error%2Cinterval%3Aauto%2C%2Csort%3A!(TimeStamp%2Cdesc)))

**Expected behavior**
Expected not to have the last `)` stripped from href.

**Comment by teimurjan** [comment]
The problem caused by the usage of [this regex](https://github.com/markedjs/marked/blob/master/lib/marked.js#L616). 

[Testing the Regex](https://regex101.com/r/N7c7Vx/1/).

**Comment by teimurjan** [comment]
If someone could explain me what is the purpose of this regexp, I'd fix the issue by myself

**Comment by UziTech** [comment]
The purpose of that is to prevent urls from include punctuation.

e.g.
```
(https://example.com)
My website is https://example.com.
https://example1.com, https://example2.com, https://example3.com
```
`).,` in those examples should not be part of the url

It is very difficult to determine what should be and what should not be part of a url.

```
My website is https://example.com?web=site!
```

Should `!` be included in the url? Maybe it is part of the query? There is no way of knowing which is why common mark only includes urls surrounded by `<>` as links.

**Comment by UziTech** [comment]
The fix would be to wrap your urls with `<>` [demo](https://marked.js.org/demo/?outputType=html&text=%3Chttps%3A%2F%2Ftest.test%3F_g%3D(refreshInterval%3A(display%3AOff%2Cpause%3A!f%2Cvalue%3A0)%2Ctime%3A(from%3A%272015-10-11T11%3A08%3A59.548Z%27%2Cmode%3Aabsolute%2Cto%3A%272015-10-11T11%3A23%3A59.548Z%27))%26_a%3D(columns%3A!(URL)%2Cindex%3Alogs_web_error%2Cinterval%3Aauto%2C%2Csort%3A!(TimeStamp%2Cdesc))%3E&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tables%22%3A%20true%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)

**Comment by teimurjan** [comment]
@UziTech Thanks, got it. But to be absolutely calm, why is it working at CommonMark?

**Comment by UziTech** [comment]
It's not working on your [CommonMark Demo](https://spec.commonmark.org/dingus/?text=https%3A%2F%2Ftest.test%3F_g%3D(refreshInterval%3A(display%3AOff%2Cpause%3A!f%2Cvalue%3A0)%2Ctime%3A(from%3A%272015-10-11T11%3A08%3A59.548Z%27%2Cmode%3Aabsolute%2Cto%3A%272015-10-11T11%3A23%3A59.548Z%27))%26_a%3D(columns%3A!(URL)%2Cindex%3Alogs_web_error%2Cinterval%3Aauto%2C%2Csort%3A!(TimeStamp%2Cdesc))). If you view the html you will see that it doesn't set it as a link because [autolinking](https://github.github.com/gfm/#autolinks-extension-) is in the gfm spec not the commonmark spec.

**Comment by UziTech** [comment]
To be clear, this is actually a bug in marked because matched parenthesis should be part of the link but just changing the `_backpedal` regex isn't going to fix that. We should be using `findClosingBracket` similar to [links](https://github.com/markedjs/marked/blob/master/lib/marked.js#L722)

**Comment by tmorehouse** [comment]
I'm seeing similar issues when we convert our changelog to HTML (we are using standard-version to genreate the changelog):

```
(fixes [#3025](https://github.com/bah/blah/issues/3025)) ([#3029](https://github.com/bah/blah/pull/3029)) ([ad57e8c](https://github.com/bah/blah/commit/ad57e8c))
```

becomes

```
(fixes #3025 (#3029 (ad57e8c)
```

**Comment by UziTech** [comment]
@tmorehouse Your case is a little bit different since it is missing a parenthesis around the link not inside the link.

I create PR #1509 to fix this
